feat(python): Deprecate `with_column`

zundertj opened the pull request:
See #6117. As you can see, this is a big change.

TODO's:
- [x] There are two Python tests failing, as it seems that the behaviour between `with_column` and `with_columns` is different for `Series`. I need to look into this.
- [x] tests for the deprecation warning being raised have to be added
- [x] I haven't touched the Rust api yet, would we want to do the same there?

Probably later this week done.

ritchie46 commented:
>  I haven't touched the Rust api yet, would we want to do the same there?

In rust it is very common to have similar named methods for different arguments. One takes an expression the other takes a sequence of expressions. I think we should leave it like that.

zundertj commented:
Ok, the failing tests were passing in a `pl.Series` into `LazyFrame.with_columns`, whilst that was fine in `DataFrame.with_columns`. Reason is that `DataFrame.with_columns was parsing it as a literal, and then passed it on to `LazyFrame.with_columns`. I have moved that logic inside `LazyFrame.with_columns`, added a unit test to verify and now `DataFrame.with_columns is the minimum possible wrapper around `LazyFrame.with_columns`, so there should be no inconsistency.

As a side note, I see that we handle things differently for args vs named_args in `with_columns`, where the latter has more flexibility (or unclear api), but lets leave that to a separate PR.

zundertj commented on py-polars/polars/internals/lazyframe/frame.py:
```diff
@@ -2473,11 +2474,15 @@ def with_columns(
-        exprs = (
-            []
-            if exprs is None
-            else ([exprs] if isinstance(exprs, pli.Expr) else list(exprs))
```
Sorry, my brain cant handle this on a Friday night...

zundertj commented on py-polars/polars/internals/dataframe/frame.py:
```diff
@@ -5568,8 +5577,6 @@ def with_columns(
 
         """
-        if exprs is not None and not isinstance(exprs, Sequence):
-            exprs = [exprs]
```
This is now gone

ritchie46 commented:
Is everybody good to go on this one? I don't want to merge this immediately, but I want to add the deprecation on the breaking release I plan end of this month. This is core functionality IMO and I think we are becoming used enough that we should introduce a bit of a grace period.

zundertj commented:
This will cause for many users deprecation warnings, so please double/triple check that we are happy with this indeed.

alexander-beedie commented:
> This will cause for many users deprecation warnings, so please double/triple check that we are happy with this indeed.

Looks like a winner to me...

One thing: I'd make the `DeprecationWarning` text _"This method will be removed in the future"_ more specific, eg: _"This method will be removed **shortly**"_ (and maybe mention "in 0.16"?)

stinodego commented:
I think this is a good move!

And we don't have to remove `with_column` in `0.16.0`, right? As it is widely used, I would suggest to just leave it in until `0.17.0`.

I'll add one comment about the DeprecationWarining, which can be a bit more informative, I think.

stinodego commented on py-polars/polars/internals/dataframe/frame.py:
```diff
@@ -4252,7 +4256,12 @@ def with_column(self, column: pli.Series | pli.Expr) -> DataFrame:
+        warnings.warn(
+            "`with_column` has been deprecated in favor of `with_columns`."
+            " This method will be removed in the future",
+            category=DeprecationWarning,
+        )
```
Add `stacklevel=2` so that people can see which line of their code raises the warning.

```suggestion
            category=DeprecationWarning,
            stacklevel=2,
        )
```

stinodego commented on py-polars/polars/internals/lazyframe/frame.py:
```diff
@@ -2613,6 +2621,12 @@ def with_column(self: LDF, column: pli.Series | pli.Expr) -> LDF:
+        warnings.warn(
+            "`with_column` has been deprecated in favor of `with_columns`."
+            " This method will be removed in the future",
+            category=DeprecationWarning,
+        )
```
```suggestion
            category=DeprecationWarning,
            stacklevel=2,
        )
```

stinodego commented on py-polars/polars/internals/dataframe/frame.py:
```diff
@@ -4215,6 +4216,9 @@ def with_column(self, column: pli.Series | pli.Expr) -> DataFrame:
         existing data.
 
+        .. deprecated:: 0.15.14
+            `with_column` will be removed in favour of the more generic `with_columns`.
```
Deprecation warning uses "favor" (American spelling), let's be consistent in this deprecation message.

```suggestion
            `with_column` will be removed in favor of the more generic `with_columns`.
```

zundertj replied:
Good point, fixed.

stinodego commented on py-polars/polars/internals/dataframe/frame.py:
```diff
@@ -4239,7 +4243,7 @@ def with_column(self, column: pli.Series | pli.Expr) -> DataFrame:
         │ 5   ┆ 6   ┆ 36.0      │
         └─────┴─────┴───────────┘
-        >>> df.with_column(pl.col("a") ** 2)  # replaced
+        >>> df.with_columns(pl.col("a") ** 2)  # replaced
```
Can we add these doc examples to `with_columns`? I want it to be obvious that `with_columns` also works when passing a single Series / Expression. That page will probably get a lot of traffic after we release this change!

We can remove the examples here - if you want to keep them, I'd suggest using `with_column` rather than `with_columns`, because technically you're giving doc examples of an entirely different method.

zundertj replied:
I have removed the examples in the `with_column` docstring, so it is even more clear we dont want to use this anymore. The warning in the docstring + the code warning make that clear now.

I have added, using the dataframe as used in `with_columns`, examples of passing in a single expression, either with a new name or an existing name. That was missing from `with_columns`, and is important for users to know.

stinodego replied:
Great!

zundertj commented:
Rebased & squashed.

zundertj commented:
> I think this is a good move!
> 
> And we don't have to remove `with_column` in `0.16.0`, right? As it is widely used, I would suggest to just leave it in until `0.17.0`.
> 
> I agree with @alexander-beedie that we should include this version information in the deprecation message, to help users schedule their upgrade.
> 
> Also, I am going to rebrand this as an 'enhancement', since this affects our public API.

I had on purpose not added in a version, because we don't know (yet). I think, given that the warnings will be annoying, people would update relatively quickly their code after the Polars upgrade, but we will have to cater to users who update irregularly. So I did not want to pin us down yet on a timeline. But if there is a good idea for how long it will be until 0.16 or 0.17 land, we can put that in.

stinodego commented:
Fair enough. Maybe @ritchie46 can comment on this. Once we decide on what the deprecation message should be, this PR should be good to go!

ritchie46 commented:
Just to be certain, we now accept 

* a single expression
* a list of expressions
* multiple expressions as args

in `with_columns` right?

stinodego commented:
Yes, that's correct!

ritchie46 commented:
Alright. Thanks all!

stinodego commented:
I'd say `0.17.0` is fine, as it's still quite far away.

zundertj commented:
Added 0.17 to docstring and the warning.

stinodego commented:
@ritchie46 is there still some doubt or unclarity about this change?

If not, I think we should go ahead and merge this, as it's big and very prone to merge conflicts.

ritchie46 commented:
Alright! Thanks all!

timreuscher commented:
Hadn't updated since 0.14.18, just installed 0.15.18, ran some code, saw the warning which made complete sense in both the warning displayed and the change it warned about, did a few search-and-replaces (I guess that's the plural) and no more warnings in less than five minutes. 

Just wanted to leave my two cents as an occasional user out in the field that this is an excellent way to deprecate old methods. Well done!

stinodego commented:
Thanks! Always nice to hear some positive feedback 😸
